arraycontext/parameter_study/__init__.py

```diff
@@ -0,0 +1,336 @@
+        placeholder_args = [_get_f_placeholder_args_for_param_study(arg, iarg,
+                                        input_id_to_name_in_program, self.actx)
+                            for iarg, arg in enumerate(args)]
+        output_template = self.f(*placeholder_args,
+                **{kw: _get_f_placeholder_args_for_param_study(arg, kw,
+                                               input_id_to_name_in_program,
+                                               self.actx)
+                    for kw, arg in kwargs.items()})
+
+        self.actx._compile_trace_callback(self.f, "post_trace", output_template)
+
+        if (not (is_array_container_type(output_template.__class__)
+                 or isinstance(output_template, Array))):
+            # TODO: We could possibly just short-circuit this interface if the
+            # returned type is a scalar. Not sure if it's worth it though.
+            raise NotImplementedError(
+                f"Function '{self.f.__name__}' to be compiled "
+                "did not return an array container or pt.Array,"
+                f" but an instance of '{output_template.__class__}' instead.")
+
+        def _as_dict_of_named_arrays(keys, ary):
+            name = "_pt_out_" + _ary_container_key_stringifier(keys)
+            output_id_to_name_in_program[keys] = name
+            dict_of_named_arrays[name] = ary
+            return ary
+
+        rec_keyed_map_array_container(_as_dict_of_named_arrays,
+                                      output_template)
+
+        placeholder_name_to_parameter_studies: dict[str, StudiesT] = {}
+        for key, val in arg_id_to_descr.items():
+            if isinstance(val, LeafArrayDescriptor):
+                name = input_id_to_name_in_program[key]
+                for axis in arg_id_to_arg[key].axes:
+                    tags = axis.tags_of_type(ParameterStudyAxisTag)
+                    if tags:
+                        if name in placeholder_name_to_parameter_studies.keys():
+                            placeholder_name_to_parameter_studies[name].append(tags)
+
+                        else:
+                            placeholder_name_to_parameter_studies[name] = tags
+
+        breakpoint()
+        expand_map = ExpansionMapper(placeholder_name_to_parameter_studies)
+        # Get the dependencies
+
+        sing_inst_outs = make_dict_of_named_arrays(dict_of_named_arrays)
+
+        # Use the normal compiler now.
+
+        compiled_func = self._dag_to_compiled_func(expand_map(sing_inst_outs),
+                                                   # pt_dict_of_named_arrays,
+                input_id_to_name_in_program=input_id_to_name_in_program,
+                output_id_to_name_in_program=output_id_to_name_in_program,
+                output_template=output_template)
+
+        breakpoint()
+        self.program_cache[arg_id_to_descr] = compiled_func
+        return compiled_func(arg_id_to_arg)
+
+
+def _cut_to_single_instance_size(name, arg) -> Array:
```

Comment:
Just build the right-size `Placeholder` to start with.

Reply:
We only build the placeholder once. I am allowing the multiple-instance argument to be passed to the function `_to_input_for_compiled(arg, actx)` whose functionality depends on the input `ArrayContext` on account of a subsequent call to `actx.transform_dag(arg)`. We are cutting to a placeholder afterwards so that the `pytato.transfrom.materialize_with_mpms` materialization strategy is still used. Other `ArrayContext`s may have other strategies implemented before the compilation step.